BUG: new 64 bit VR type - fixed Win issues (failed build, overflow)

issakomi opened the pull request:
The idea is to use for VR _long long_ instead _long_  (previous VR update in #85 _int_ to _long_ does not work on many platforms, size of _long_ is varying and can be easily same as _int_ type before), also tell compiler about VR's type with _enum:long long_  (C++11).

File gdcmTagToType.h have to be re-generated, xls file was updated

**Edit 2:** done so far, note: many lines are from generated file

**Edit 1:** VS 2013 x64 compiler errors:
`12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(84): warning C4309: 'initializing' : truncation of constant value 12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(93): warning C4309: 'initializing' : truncation of constant value 12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(313): error C2766: explicit specialization; 'gdcm::VRToEncoding<0>' has already been defined 12> c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(304) : see previous definition of 'VRToEncoding<0>' 12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(313): error C2766: explicit specialization; 'gdcm::VRToType<0>' has already been defined 12> c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(304) : see previous definition of 'VRToType<0>' 12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(356): error C2196: case value '0' already used 12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(360): error C2196: case value '0' already used 12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(369): error C2196: case value '1' already used 12> gdcmDataSet.cxx 12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(84): warning C4309: 'initializing' : truncation of constant value 12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(93): warning C4309: 'initializing' : truncation of constant value 12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(313): error C2766: explicit specialization; 'gdcm::VRToEncoding<0>' has already been defined 12> c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(304) : see previous definition of 'VRToEncoding<0>' 12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(313): error C2766: explicit specialization; 'gdcm::VRToType<0>' has already been defined 12> c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(304) : see previous definition of 'VRToType<0>' 12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(356): error C2196: case value '0' already used 12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(360): error C2196: case value '0' already used 12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(369): error C2196: case value '1' already used 12> gdcmExplicitDataElement.cxx

<...>`

issakomi commented:
Re-created with updated TagToType.xsl
xsltproc TagToType.xsl Part6.xml > gdcmTagToType.h

issakomi commented:
Created with the branch-   [ConformVR.dcm](https://drive.google.com/file/d/1L5Bi1oK3yW_z6KntSZjh46fVMHEmxjMr/view?usp=sharing)

gdcmdump (after update)

signed long min, max
unsigned long min, max
```
(4d4d,1030) SV -9223372036854775808\9223372036854775807         # 16,? (2)  GDCM:UNKNOWN
(4d4d,1031) UV 0\18446744073709551615                         # 16,? (2)  GDCM:UNKNOWN
```

issakomi commented:
@malaterre 
i am still thinking about that huge generated _gdcmTagToType.h_
`xsltproc TagToType.xsl Part6.xml > gdcmTagToType.h`
May be better to edit the old one, it is easy
enum { VRType = ... } --------> enum:long long { VRType = ... }

What do you prefer?

malaterre commented:
@issakomi Technically you should get the *exact* same result so I would use `sed` first, then tweak xsl file and check files are identical.

issakomi commented:
@malaterre 
I've taken old gdcmTagToType.h file (only updated related _enum_). Generated file was too different, i think it is better not too change it now, may be something will go out of sync or whatever.

issakomi commented:
Tested win32 and win64 builds, generated test files (same as ConformVR.dcm):
[win32](https://drive.google.com/file/d/1m2Twg-oTQCN2gkrxVLbhFAmY-bbvsCnG/view?usp=sharing)
[win64](https://drive.google.com/file/d/1HZ6kQC516a8SFa8z7vDCcBvL9IUnltVY/view?usp=sharing)
So far done for now.

malaterre commented:
@issakomi I fail to understand the `enum:long long` portion of the patch. Are you sure this is required, it seems to me enum {} should be enough in this case, since compiler is required to handle all values listed in the `enum{}` itself. I wont argue much about it, but I am curious as to why.

malaterre commented:
Also feel free to rebase + squash your commits (followed by git push -f) to keep the history nice and clean :) thanks

malaterre commented on Source/DataStructureAndEncodingDefinition/gdcmAttribute.h:
```diff
@@ -79,14 +79,14 @@ template<> class VRVLSize<1> {
 public:
   typedef typename VRToType<TVR>::Type ArrayType;
-  enum { VMType = VMToLength<TVM>::Length };
+  enum:long long { VMType = VMToLength<TVM>::Length };
```
Why does this patch touches `VM`, I assumed this should only touches `VR` stuff

issakomi replied:
`enum:long long { VMType = VMToLength<TVM>::Length };`
is probably not necessary here (2 places), not sure

issakomi replied:
@malaterre 
should i remove it? will probably work without, IMHO. But can not make Windows tests just now.

issakomi replied:
@malaterre
Done revert VMType change

issakomi commented:
it is new C++11 thingy, without it i couldn't get rid of truncation on Windows (was always int32).
With Linux absolutely no problem, _long_ and _long long_ seem to be always int64 and _enum_ type too,
this code will compile on Linux without C++11 too (with harmless warning)
[Documentation - enum:long long](https://en.cppreference.com/w/cpp/language/enum)

malaterre commented:
@issakomi I am not arguing about the c++11 notation, I am arguing about your compiler failing to find the corresponding underlying type.

Eg:

* https://en.cppreference.com/w/cpp/language/enum

> 
Values of unscoped enumeration type are implicitly-convertible to integral types. If the underlying type is not fixed, the value is convertible to the first type from the following list able to hold their entire value range: int, unsigned int, long, unsigned long, long long, or unsigned long long.

issakomi commented:
@malaterre 
I've done minimal changes: changed return type of VR::GetIndex from _long long_ to _unsigned int_.
It was _long_ before, so i initially took _long long_, but i realized that it is not required at all, it is usual index function like VM::GetIndex and should better return unsigned. This also fixed one warning.
I think that i am ready with 64 bit VR issue. Please kindly allow me to skip that rebase-squish-force thingy, i am not experienced with. Thank you.

issakomi commented:
@malaterre 
i've read just now you are going to fix or revert OpenJPEG commit. If so i can wait and merge with upstream, it think it will be easier.

malaterre commented:
@issakomi this is merged in master, only left to do is Source/DataDictionary/TagToType.xsl. I'll handle it in a second step. Thanks much for the bug fix!

issakomi commented:
@malaterre 
Thank you. I hope everything is OK with TagToType.xsl, i tried to re-create many files with xsltproc, they are mostly little different, thought that may be Part6.xml has changed, not sure. Current gdcmTagToType.h is the old one, but manually edited (only _emums_).

issakomi commented:
@malaterre 
Unfortunately important things are gone again, that enum:long long was required, it is disappered from gdcmVR.h and TagToType, don't know what else changed. Master doesn't complile again (tried only VS 2013 32 bit now).

```
12>  gdcmDataElement.cxx
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(84): warning C4309: 'initializing' : truncation of constant value
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(93): warning C4309: 'initializing' : truncation of constant value
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(313): error C2766: explicit specialization; 'gdcm::VRToEncoding<0x00>' has already been defined
12>          c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(304) : see previous definition of 'VRToEncoding<0>'
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(313): error C2766: explicit specialization; 'gdcm::VRToType<0x00>' has already been defined
12>          c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(304) : see previous definition of 'VRToType<0>'
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(356): error C2196: case value '0' already used
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(360): error C2196: case value '0' already used
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(369): error C2196: case value '1' already used
12>  gdcmDataSet.cxx
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(84): warning C4309: 'initializing' : truncation of constant value
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(93): warning C4309: 'initializing' : truncation of constant value
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(313): error C2766: explicit specialization; 'gdcm::VRToEncoding<0x00>' has already been defined
12>          c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(304) : see previous definition of 'VRToEncoding<0>'
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(313): error C2766: explicit specialization; 'gdcm::VRToType<0x00>' has already been defined
12>          c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(304) : see previous definition of 'VRToType<0>'
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(356): error C2196: case value '0' already used
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(360): error C2196: case value '0' already used
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(369): error C2196: case value '1' already used
12>  gdcmExplicitDataElement.cxx
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(84): warning C4309: 'initializing' : truncation of constant value
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(93): warning C4309: 'initializing' : truncation of constant value
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(313): error C2766: explicit specialization; 'gdcm::VRToEncoding<0x00>' has already been defined
12>          c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(304) : see previous definition of 'VRToEncoding<0>'
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(313): error C2766: explicit specialization; 'gdcm::VRToType<0x00>' has already been defined
12>          c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(304) : see previous definition of 'VRToType<0>'
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(356): error C2196: case value '0' already used
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(360): error C2196: case value '0' already used
12>c:\users\mi\desktop\b\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(369): error C2196: case value '1' already used
```

malaterre commented:
@issakomi Send me the output of:

`bin/gdcmDSEDTests "TestVR"`

using latest gdcm/master using your compiler (you may want to use your long long trick to get passed compilation error, thanks much).

In particular I am interested in:

* https://github.com/malaterre/GDCM/commit/e2dbdbeca6402278a2a8eb36eeeed51d95c78662

issakomi commented:
Hmm...  i can not compile on Windows because it doesn't compile, i recommend to test the issue on Windows (64 or 32 bit, i have only VS 2013 C++), seems on your platform the problems are not visible, i understand, i've always overseen them, because i use usually Linux.

issakomi commented:
Tried this code:

```
	std::cout << "size of int        "
		<< sizeof(int) << std::endl;

	std::cout << "size of long       "
		<< sizeof(long) << std::endl;

	std::cout << "size of long long  "
		<< sizeof(long long) << std::endl;

	enum { V1 = (int)INT_MAX } E1;
	std::cout << "size of enum { INT_MAX  }    "
		<< sizeof(E1) << std::endl;

	enum { V2 = (unsigned int)UINT_MAX } E2;
	std::cout << "size of enum { UINT_MAX }    "
		<< sizeof(E2) << std::endl;

	enum { V3 = (long long)LLONG_MAX } E3;
	std::cout << "size of enum { LLONG_MAX }   "
		<< sizeof(E3) << std::endl;

	enum { V4 = (unsigned long long)ULLONG_MAX } E4;
	std::cout << "size of enum { ULLONG_MAX }  "
		<< sizeof(E4) << std::endl;

	enum { V5 = 1 } E5;
	std::cout << "size of enum { 1 }           "
		<< sizeof(E5) << std::endl;

	enum:long long { V6 = 1 } E6;
	std::cout << "size of enum:long long { 1 } "
		<< sizeof(E6) << std::endl;
```

Output on win64 compiled with VS 2013 C++ 64 bit:

```
C:\Users\mi\Desktop\test009\build\Release>test.exe
size of int        4
size of long       4
size of long long  8
size of enum { INT_MAX  }    4
size of enum { UINT_MAX }    4
size of enum { LLONG_MAX }   4
size of enum { ULLONG_MAX }  4
size of enum { 1 }           4
size of enum:long long { 1 } 8
```

agravgaard commented:
The Intel compiler (Version 19.0.3) on Windows 64bit is not able to determine the appropriate type for the enum either.
Changing to `enum : long long {` in `gdcmVR.h` and `gdcmTagToType.h` fixes the problem for me.
Otherwise `UV` and `SV` will both be initialized to `0`, causing an error, while `VR::VRALL` will only cause a warning.
Example with MSVC based on `gdcmVR.h`: https://godbolt.org/z/uuUmig
